tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -7035,28 +7043,60 @@ void CGMSHLSLRuntime::EmitHLSLOutParamConversionInit(
         } else {
           continue;
         }
+      } else if (isAggregateType) {
+        // aggregate in-only - emit RValue, unless LValueToRValue cast
+        EmitRValueAgg = true;
+        if (const ImplicitCastExpr *cast =
+                dyn_cast<ImplicitCastExpr>(Arg)) {
+          if (cast->getCastKind() == CastKind::CK_LValueToRValue) {
+            EmitRValueAgg = false;
+          }
+        }
+      } else {
+        // Must be object
+        DXASSERT(isObject, "otherwise, flow condition changed, breaking assumption");
+        // in-only objects should be skipped to preserve previous behavior.
+        continue;
       }
     }
 
-    // get original arg
-    LValue argLV = CGF.EmitLValue(Arg);
+    // Skip unbounded array, since we cannot preserve copy-in copy-out
```

Comment:
Are unbounded arrays an exception that is passed by pointer or are they simply not supported (which would make more sense)? If the latter, an assert would be more appropriate.

Reply:
We need to do some work here to support these in whatever way we think makes sense, but there's a test that will exercise this case, and asserting here is likely to be a regression.  Fxc supports certain cases without copy-in/copy-out behavior, for shader model 5.1.  Some of it's support, however, is a bit contradictory between SM 5.0 and 5.1, so we have to sort out what we need to support, and how to clearly specify that.  I was not trying to solve this problem in this change, but had to do something to not introduce a regression here.